File: data_structures/queue/circular_queue/cpp/circular queue.cpp
#include<iostream>
#include<conio.h>
using namespace std;

class queue1
{

    int n=8,front1=-1,rear=-1,a=0;
public:

    int arr[8];

    void element()
    {

        cout<<"enter element";
        cin>>a;
    }
    void inqueue()
    {
    if((front1>=0&&rear==n-1)||(front1==rear+1))
            {
             //overflow
             cout<<"queue overflow"<<endl;
            }
   if (front1==-1&&rear==-1)
        {
            //queue empty
            front1=0;
            rear=0;
            arr[rear]=a;

        }
        else if(rear==n-1)
            {
                rear=0;
                arr[rear]=a;

            }

        else {
                rear++;
               arr[rear]=a;
            }
    }
    int b;
    void dequeue()
    {
        if(rear==-1&&front1==-1)
        {
            cout<<"underflow"<<endl;
        }
    if (front1==rear)
        {
            //one element
            b=arr[front1];
            cout<<"deleted single element is :"<<b<<endl;
            front1=-1;
            rear=-1;

        }
        else if (front1==n-1)
        {
            b=arr[front1];
            cout<<"deleted element last is :"<<b<<endl;
            front1=0;
        }
        else
        {
           b=arr[front1];
            cout<<"deleted any element is :"<<b<<endl;
            front1++;
        }

    }
    void display()
    {
        cout<<"the no. in queue are"<<endl;
        for(int i=front1;i<=rear;i++)
        {
            cout<<arr[i]<<endl;
        }
    }
};
    int main()
    {
        queue1 q;
        int ch,c=1;

while(c==1)
{
        cout<<"enter choice 1.insertion 2.deletion 3.display 4.exit"<<endl;
        cin>>ch;
   switch(ch)
{
    case 1:
        {
            q.element();
            q.inqueue();
            break;
        }
    case 2:
        {
            q.dequeue();
            break;
        }
    case 3:
        {
            q.display();
            break;
        }
    case 4:
        {
            c=0;
            break;
        }
}
}
}


//seraching n traversing in lkd list

File: data_structures/circular_queue/java/circular_queue.java
//circular queue
import java.util.*;
interface queue1
{
	void insert();
	void delete();
	void display();
	
}
class myqueue1 implements queue1
{
	int arr[]=new int[3];
	int front=0,rear=0;
	Scanner s1 = new Scanner(System.in);
	public void insert()
	{
			int a;
			System.out.println("enter element:");
			a=s1.nextInt();
			arr[rear]=a;
			rear++;
		if(rear==3)
			rear=0;
	}
public void delete()
	{
		if (rear==front)
			System.out.println("queue is empty");
		else
		{
			int a;
			a=arr[front];
			front++;
			System.out.println("element deleted : "+a );
			if(front==3)
			front=0;
		}
	}
	public void display()
	{
		System.out.println("elements are");
		for(int i=front;i<rear;i++)
		{
			System.out.println(arr[i]);
		}
	}
}
class tut7
{
public static void main(String args[])
{
	Scanner s = new Scanner(System.in);
		myqueue1 m = new myqueue1();
		boolean b= true;
		while(b)
		{
		System.out.println("enter ur choices \n 1. insert \n 2. delete \n 3. display \n 4. exit");
		int n;
		n=s.nextInt();
		
		
		switch(n)
		{
		case 1:
		m.insert();
		break;
		case 2:
		m.delete();
		break;
		case 3:
		m.display();
		break;
		case 4:
		System.exit(0);
		}
		}
	}
}
